### PR TITLE
Adds logic to retain undo/redo history for TextInput

### LINKED ISSUE #10132: Undo / redo history is lost when controlled prop updates occur on TextInput
### Problem Description

When building a component with controlled prop updates to TextInput (XAML TextBox), any prop updates from JS cause the TextBox to lose it's undo / redo history. While this is sometimes appropriate, it may be more useful in general if the controlled prop updates can retain history.

For example, consider a TextInput that handles mentions via JS logic and controlled prop updates. After selecting a mention, it should be feasible for the user to undo the mention selection.

### Steps To Reproduce

1. Run RNTester > TextInput example
2. In one of the "Live Re-Write" examples, insert text into the TextInput that would cause a live re-write / controlled prop update.
3. Notice that there is no undo capability after the live re-write has occurred.

### Expected Results

We should be able to undo the controlled prop update from JS, and all history should not be lost.

### CLI version

npx react-native --version

### Environment

```markdown
npx react-native info
```

### Target Platform Version

_No response_

### Target Device(s)

_No response_

### Visual Studio Version

_No response_

### Build Configuration

_No response_

### Snack, code example, screenshot, or link to a repository

See that Undo option is gone due to controlled prop update:

https://user-images.githubusercontent.com/1106239/174145046-203dc0c2-8ff6-4efe-a24e-92267f6fcef3.mp4

### PR DESCRIPTION
**Pull request by rozele**
### Type of Change
_Erase all that don't apply._
- Bug fix (non-breaking change which fixes an issue)

### Why
As described in #10132, controlled text updates from JS to XAML TextBox for TextInput cause the undo redo history to be lost because imperatively setting the XAML TextBox::Text property resets history on the TextBox.

Resolves #10132

### What
This change finds a single diff in the before and after string and uses the SelectedText property which retains undo/redo history.

## Testing

https://user-images.githubusercontent.com/1106239/174148621-695de902-72be-4e28-a15e-0dd16dcee61b.mp4

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/10133)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 1]
```diff
@@ -712,9 +712,39 @@ void TextInputShadowNode::SetText(const winrt::Microsoft::ReactNative::JSValue &
...
+            while (diffEndIndex < oldValue.size() - diffStartIndex && diffEndIndex < newValue.size() - diffStartIndex &&
+                   oldValue[oldValue.size() - diffEndIndex - 1] == newValue[newValue.size() - diffEndIndex - 1])
+              diffEndIndex++;
+            // 3. Select the range between the start and end index in the "old value"
+            textBox.SelectionStart(diffStartIndex);
+            textBox.SelectionLength(oldValue.size() - diffStartIndex - diffEndIndex);
+            // 4. Replace the selected text with the range between start and end index in the "new value"
+            auto replacementValue =
+                std::wstring_view{newValue}.substr(diffStartIndex, newValue.size() - diffStartIndex - diffEndIndex);
+            textBox.SelectedText(replacementValue);
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:737]: This occasionally crashes - likely due to move semantics on the string view.

**Comment by rozele** [comment]
Would this be more acceptable with a quirk setting? cc @chrisglein @asklar

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 2]
```diff
@@ -712,9 +712,40 @@ void TextInputShadowNode::SetText(const winrt::Microsoft::ReactNative::JSValue &
...
+            // undo / redo history. To do so, we use a simple algorithm that finds a
+            // single diff in the values:
+            // 1. Find first character that mismatches iterating forwards
+            while (diffStartIndex < oldValue.size() && diffStartIndex < newValue.size() &&
+                   oldValue[diffStartIndex] == newValue[diffStartIndex])
+              diffStartIndex++;
+            // 2. Find last character the mismatches beyond the first mismatch iterating backwards
+            while (diffEndIndex < oldValue.size() - diffStartIndex && diffEndIndex < newValue.size() - diffStartIndex &&
+                   oldValue[oldValue.size() - diffEndIndex - 1] == newValue[newValue.size() - diffEndIndex - 1])
+              diffEndIndex++;
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:730]: Would be good to have test cases for this to ensure that this produces the right results for different string values. Seems like something unit testable on its own (no need for UI automation), and also would benefit from UI automation. There might not be the right infrastructure for that now, but if it were that'd be the right call here.

Also feels like this block should be a named function/lambda for clarity ("findUnchangedRange" or whatever)

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 3]
```diff
@@ -712,9 +712,40 @@ void TextInputShadowNode::SetText(const winrt::Microsoft::ReactNative::JSValue &
...
+            uint32_t diffEndIndex = 0;
+            // Replacing the entire string with the new value resets the text history.
+            // Generally speaking, this is undesirable. To mitigate this issue, we can
+            // replace text using the XAML TextBox::SelectedText property, which retains
+            // undo / redo history. To do so, we use a simple algorithm that finds a
+            // single diff in the values:
+            // 1. Find first character that mismatches iterating forwards
+            while (diffStartIndex < oldValue.size() && diffStartIndex < newValue.size() &&
+                   oldValue[diffStartIndex] == newValue[diffStartIndex])
+              diffStartIndex++;
```
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:726]:
```suggestion
            while (diffStartIndex < oldValue.size() && diffStartIndex < newValue.size() &&
                   oldValue[diffStartIndex] == newValue[diffStartIndex]) {
              diffStartIndex++;
            }
```

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp (hunk 2 above)
Comment (chrisglein) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:730]:
```suggestion
            while (diffEndIndex < oldValue.size() - diffStartIndex && diffEndIndex < newValue.size() - diffStartIndex &&
                   oldValue[oldValue.size() - diffEndIndex - 1] == newValue[newValue.size() - diffEndIndex - 1]) {
              diffEndIndex++;
            }
```

**Review by chiaramooney — APPROVED** [review]
👍

**Pull request merged**